src/main/java/com/openlattice/chronicle/services/ChronicleServiceImpl.java

```diff
@@ -215,61 +230,88 @@ private UUID reserveEntityKeyId(
     }
 
     private UUID reserveDeviceEntityKeyId(
+            UUID devicesESID,
             Map<UUID, Set<Object>> data,
             DataIntegrationApi dataIntegrationApi ) {
 
         return reserveEntityKeyId(
-                entitySetIdMap.get( DEVICES_ENTITY_SET_NAME ),
+                devicesESID,
                 ImmutableList.of( propertyTypeIdsByFQN.get( STRING_ID_FQN ) ),
                 data,
                 dataIntegrationApi
         );
     }
 
-    private UUID getStudyEntityKeyId( UUID studyId ) {
-        logger.info( "Retrieving studyEntityKeyId, studyId = {}", studyId );
-        if ( studies.containsKey( studyId ) ) {
-            return studies.get( studyId );
-        }
+    private UUID getStudyEntityKeyId( UUID organizationId, UUID studyId ) {
+        logger.info( "Retrieving studyEKID: studyId = {}, orgId = {}", studyId, organizationId );
+        try {
+            // api
+            ApiClient apiClient = prodApiClientCache.get( ApiClient.class );
+            DataApi dataApi = apiClient.getDataApi();
 
-        logger.error( "Failed to retrieve studyEntityKeyId, studyId = {}", studyId );
-        return null;
+            UUID entitySetId = ensureEntitySetExists( organizationId, CHRONICLE_CORE, STUDIES );
+
+            String jwtToken = auth0Client.getIdToken( username, password );
+
+            Iterable<SetMultimap<FullQualifiedName, Object>> data = dataApi
+                    .loadEntitySetData( entitySetId, FileType.json, jwtToken );
+
+            return StreamUtil.stream( data )
+                    .filter( entry -> !entry.get( STRING_ID_FQN ).isEmpty() )
+                    .filter( entry -> entry.get( STRING_ID_FQN ).iterator().next().toString()
```

Comment:
since `entry.get( someFqn ).iterator().next().toString()` is such a common pattern, I think it would clean up the code a lot to give it its own private helper function (and also the UUID version). i.e.

```
private String getFirstValueOrNull( SetMultimap <FullQualifiedName, Object> entity, FullQualifiedName fqn ) {
  if ( entity.get( fqn ).isEmpty() ) {
    return null;
  }

  return entity.get( fqn ).iterator().next().toString();
}

private String getFirstUUIDOrNull( SetMultimap <FullQualifiedName, Object> entity, FullQualifiedName fqn ) {
  String firstValue = getFirstValueOrNull( entity, fqn );
  if ( firstValue == null ) {
    return null;
  }

  return UUID.fromString( firstValue );
}
```

^ this also heavily simplifies the filtering step, which could then come after the mapping and could just check for nulls. So it would become something like 
```
  return StreamUtil.stream( data )
    .map( entry -> getFirstUUIDOrNull( entry, STRING_ID_FQN )
    .filter( id -> id != null )
    .findFirst().orElse( null )
```

Reply:
Great suggestions! Thanks